docs: add doc on kuberentes sync manifests (#515)
siderolabs/docs: public/omni/cluster-management/sync-kubernetes-manifests.mdx

---
title: "Sync Kubernetes Manifests"
description: "Deploy and sync Kubernetes workloads directly from Omni cluster templates using manifest sync."
---

Omni's manifest sync feature lets you define Kubernetes workloads directly inside a [cluster template](../reference/cluster-templates).
When you apply a cluster template with manifests, Omni stores the manifest data in its state and synchronizes it to the cluster once the Kubernetes API is available and the cluster is healthy.

> This is useful for bootstrapping workloads like Argo CD, NGINX, or any other application you want running immediately after a cluster is created.

Reviewer: ```suggestion
This is useful for bootstrapping workloads like Argo CD, custom CNI, or any other application you want running immediately after a cluster is created.
```

## Prerequisites

Before you get started, you must have the following:

- An Omni account with permission to manage clusters.
- `omnictl` installed and configured. See [Install and configure omnictl](../getting-started/install-and-configure-omnictl).

## How manifest sync works

Omni pushes manifest data to its internal state and applies it to the cluster asynchronously.
Omni waits until the Kubernetes API is available and the cluster is healthy before attempting to apply manifests.
If a manifest fails to apply due to a transient error, Omni retries continuously until it succeeds.

> If a manifest contains a configuration error (for example, an invalid field), Omni reports the error in the `KubernetesManifestStatus` resource and keeps retrying.

Reviewer: ```suggestion
If a manifest contains a configuration error (for example, an invalid field), Omni reports the error in the `ClusterKubernetesManifestsStatus` resource and keeps retrying.
```

### Inline manifest

Use `inline` to embed Kubernetes resource definitions directly in the cluster template.
The `inline` field accepts a list of standard Kubernetes manifests.
All standard Kubernetes fields are supported and are forwarded to the Kubernetes client without modification.

```yaml
kind: Cluster
name: manifests
kubernetes:
  version: v1.35.0
manifests:
  - name: nginx
    mode: full
    inline:
      - apiVersion: apps/v1
        kind: Deployment
        metadata:
          name: nginx-deployment
          namespace: default
        spec:
          selector:
            matchLabels:
              app: nginx
          replicas: 2
          template:
            metadata:
              labels:
                app: nginx
            spec:
              containers:
                - name: nginx
                  image: nginx:1.28.0
                  ports:
                    - containerPort: 80
```

### External file manifest

Use `file` to reference an external YAML file.
The file path is relative to the working directory when running `omnictl`.

```yaml
kind: Cluster
name: manifests
kubernetes:
  version: v1.35.0
manifests:
  - name: another

>     file: /home/artem/workspace/omni/manifest.yaml

Reviewer: Lets not have my name here. Also please use the relative path

    mode: full
```

### Multiple manifests

You can define multiple manifest entries under the `manifests` field.
Each entry is applied independently and can use a different mode.

```yaml
kind: Cluster
name: manifests
kubernetes:
  version: v1.35.0
manifests:
  - name: nginx
    mode: full
    inline:
      - apiVersion: apps/v1
        kind: Deployment
        metadata:
          name: nginx-deployment
          namespace: default
        spec:
          selector:
            matchLabels:
              app: nginx
          replicas: 2
          template:
            metadata:
              labels:
                app: nginx
            spec:
              containers:
                - name: nginx
                  image: nginx:1.28.0
                  ports:
                    - containerPort: 80
  - name: another

>     file: /home/artem/workspace/omni/manifest.yaml

Reviewer: Same here

Author: oh an oversight on my part, will remove

    mode: full
```

A single manifest entry can also contain multiple Kubernetes resource definitions.
This is useful when a service ships multiple resources (for example, a Namespace, Deployment, and Service) as a single manifest.

## Apply the cluster template

Once the cluster template is updated with manifests, apply it using `omnictl`:

```bash
omnictl cluster template sync --file clustermanifests.yaml
```

Omni stores the manifest definitions and begins synchronizing them to the cluster.
If the cluster is not yet healthy, Omni waits until the Kubernetes API is available before applying.

### Verify manifest sync status

To inspect the sync status of manifests for a cluster, use `omnictl` to get the `KubernetesManifestStatus` resource:

```bash

> omnictl get kubernetesmanifeststatuses

Reviewer: ```suggestion
omnictl get clusterkubernetesmanifestsstatuses <cluster-name>
```
